src/lib.rs

```diff
@@ -0,0 +1,560 @@
+    /// tree.insert(String::from("hello"));
+    /// tree.insert(String::from("world!"));
+    /// assert!(tree.contains(&String::from("hello")));
+    /// assert!(!tree.contains(&String::from("cruel")));
+    /// ```
+    pub fn contains(&self, value: &T) -> bool {
+        self.get_node(value).is_ok()
+    }
+
+    /// Add the specified value to this tree.
+    ///
+    /// # Example
+    /// ```
+    /// use avl_tree::SearchTree;
+    /// let mut tree: SearchTree<String> = SearchTree::new();
+    /// tree.insert(String::from("hello"));
+    /// tree.insert(String::from("world!"));
+    /// assert!(tree.contains(&String::from("hello")));
+    /// assert!(tree.contains(&String::from("world!")));
+    /// ```
+    pub fn insert(&mut self, value: T) {
+        if let Ok(node) = self.create_node(value) {
+            self.nodes.push(node);
+            // inserted the root
+            if self.nodes.len() == 1 {
+                self.root = Some(0);
+            }
+            let last_added = self.nodes.len() - 1;
+            self.update_heights(last_added);
+            let mut node = Some(last_added);
+            while let Some(index) = node {
+                if !self.is_balanced(index) {
+                    self.rebalance(index);
+                }
+                node = self.nodes[index].parent;
+            }
+        }
+    }
+
+    fn get_node(&self, value: &T) -> Result<usize, Option<usize>> {
+        let mut current_index = self.root;
+        let mut previous_index = None;
+        if !self.nodes.is_empty() {
+            while let Some(index) = current_index {
+                previous_index = current_index;
+                if *value > self.nodes[index].value {
+                    current_index = self.nodes[index].right_child;
+                } else if *value < self.nodes[index].value {
+                    current_index = self.nodes[index].left_child;
+                } else {
+                    return Ok(index);
+                }
+            }
+        }
+        Err(previous_index)
+    }
+
+    fn create_node(&mut self, value: T) -> Result<Node<T>, usize> {
+        match self.get_node(&value) {
+            Ok(index) => Err(index),
+            Err(previous_index) => {
+                if let Some(parent) = previous_index {
+                    if self.nodes[parent].value < value {
+                        self.nodes[parent].right_child = Some(self.nodes.len());
+                    } else {
+                        self.nodes[parent].left_child = Some(self.nodes.len());
+                    }
+                }
+                Ok(Node {
+                    value: value,
+                    parent: previous_index,
+                    left_child: None,
+                    right_child: None,
+                    height: 0
+                })
+            }
+        }
+    }
+
+    fn update_heights(&mut self, start_index: usize) {
+        if !self.nodes.is_empty() {
```

Comment:
Similarly, is this needed?

Reply:
It is, since I call `self.root.unwrap()` further down